packages/node/src/indexer/dictionary/v2/ethDictionaryV2.ts

```diff
@@ -0,0 +1,288 @@
+            dictionaryConditions.logs.push(
+              eventFilterToDictionaryCondition(filter, ds.options),
+            );
+          } else {
+            // do nothing;
+          }
+          break;
+        }
+        default:
+      }
+    }
+  }
+
+  if (!dictionaryConditions.logs.length) {
+    delete dictionaryConditions.logs;
+  }
+
+  if (!dictionaryConditions.transactions.length) {
+    delete dictionaryConditions.transactions;
+  }
+
+  //TODO, unique
+  return dictionaryConditions;
+  // return uniqBy(
+  //   allDictionaryConditions,
+  //   (item) =>
+  //     `${item}|${JSON.stringify(
+  //       sortBy(item.conditions, (c) => c.field),
+  //     )}`,
+  // );
+}
+
+export class EthDictionaryV2 extends DictionaryV2<
+  EthereumBlock,
+  SubqlDatasource,
+  EthDictionaryV2QueryEntry
+> {
+  constructor(
+    endpoint: string,
+    nodeConfig: NodeConfig,
+    eventEmitter: EventEmitter2,
+    project: SubqueryProject,
+    chainId?: string,
+  ) {
+    super(
+      endpoint,
+      chainId ?? project.network.chainId,
+      nodeConfig,
+      eventEmitter,
+    );
+  }
+
+  static async create(
+    endpoint: string,
+    nodeConfig: NodeConfig,
+    eventEmitter: EventEmitter2,
+    project: SubqueryProject,
+    chainId?: string,
+  ): Promise<EthDictionaryV2> {
+    const dictionary = new EthDictionaryV2(
+      endpoint,
+      nodeConfig,
+      eventEmitter,
+      project,
+      chainId,
+    );
+    await dictionary.init();
+    return dictionary;
+  }
+
+  buildDictionaryQueryEntries(
+    dataSources: SubqlDatasource[],
+  ): EthDictionaryV2QueryEntry {
+    const filteredDs = ethFilterDs(dataSources);
+    return buildDictionaryV2QueryEntry(filteredDs);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/require-await
+  async getData(
+    startBlock: number,
+    endBlock: number,
+    limit: number = MIN_FETCH_LIMIT,
+  ): Promise<DictionaryResponse<IBlock<EthereumBlock> | number> | undefined> {
+    return super.getData(startBlock, endBlock, limit, {
+      blockHeader: true,
+      logs: { transaction: true },
+      transactions: { log: true },
+    });
+  }
+
+  convertResponseBlocks<RFB = RawEthBlock>(
+    data: RawDictionaryResponseData<RFB>,
+  ): DictionaryResponse<IBlock<EthereumBlock>> | undefined {
+    try {
+      const blocks: IBlock<EthereumBlock>[] = [];
+      for (const block of (data as any).blocks) {
+        blocks.push(rawBlockToEthBlock(block));
+      }
+      if (blocks.length !== 0) {
+        return {
+          batchBlocks: blocks,
+          lastBufferedHeight: blocks[blocks.length - 1].block.number,
+        };
+      } else {
+        return {
+          batchBlocks: [],
+          lastBufferedHeight: undefined, // TODO is this correct?
+        };
+      }
+    } catch (e) {
+      logger.error(
+        e,
+        `Failed to handle block response ${JSON.stringify(data)}`,
+      );
+      throw e;
+    }
+  }
```

Comment:
I think this can be simplified, it would be good to have a test for an empty block response

Reply:
Added a test here https://github.com/subquery/subql/pull/2307/files#diff-eaff4dbec3da57f4f819ea6a63d9049c0727901caecbb14392bc1ea282bf64c6R148